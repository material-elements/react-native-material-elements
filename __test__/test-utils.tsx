import React from 'react';
import { render, RenderOptions } from '@testing-library/react-native';
import { ThemeProvider } from '../src';

export const ThemeWrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  return <ThemeProvider>{children}</ThemeProvider>;
};

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: ThemeWrapper, ...options });

export * from '@testing-library/react-native';
export { customRender as render };
