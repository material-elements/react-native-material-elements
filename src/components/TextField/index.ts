export * from './TextField';
export * from './BaseInput';
export * from './InputLabel';
export * from './IconInput';
export * from './OtpInput';
