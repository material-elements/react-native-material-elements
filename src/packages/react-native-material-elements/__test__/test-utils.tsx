import React from 'react';
import { render, RenderOptions } from '@testing-library/react-native';
import { ThemeProvider } from '../src';
import { ThemeProviderProps } from '../src/libraries/types';

interface CustomRenderOptions extends Omit<RenderOptions, 'wrapper'> {
  themeProps?: Partial<ThemeProviderProps>;
}

export const ThemeWrapper: React.FC<ThemeProviderProps> = ({ children, ...props }) => {
  return <ThemeProvider {...props}>{children}</ThemeProvider>;
};

export const customRender = (ui: React.ReactElement, options?: CustomRenderOptions) => {
  const { themeProps, ...rest } = options || {};
  return render(ui, {
    wrapper: wrapperProps => <ThemeWrapper {...themeProps}>{wrapperProps.children}</ThemeWrapper>,
    ...rest,
  });
};

export * from '@testing-library/react-native';
export { customRender as render };
