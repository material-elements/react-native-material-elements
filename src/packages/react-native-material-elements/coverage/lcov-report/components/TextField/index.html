
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for components/TextField</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> components/TextField</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">85.85% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>85/99</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">70.9% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>39/55</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">81.81% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>18/22</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">85.85% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>85/99</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line high'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="BaseInput.tsx"><a href="BaseInput.tsx.html">BaseInput.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	</tr>

<tr>
	<td class="file high" data-value="IconInput.tsx"><a href="IconInput.tsx.html">IconInput.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	</tr>

<tr>
	<td class="file high" data-value="InputLabel.tsx"><a href="InputLabel.tsx.html">InputLabel.tsx</a></td>
	<td data-value="84.61" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 84%"></div><div class="cover-empty" style="width: 16%"></div></div>
	</td>
	<td data-value="84.61" class="pct high">84.61%</td>
	<td data-value="13" class="abs high">11/13</td>
	<td data-value="60" class="pct medium">60%</td>
	<td data-value="5" class="abs medium">3/5</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="4" class="abs medium">3/4</td>
	<td data-value="84.61" class="pct high">84.61%</td>
	<td data-value="13" class="abs high">11/13</td>
	</tr>

<tr>
	<td class="file high" data-value="InputOutline.tsx"><a href="InputOutline.tsx.html">InputOutline.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	</tr>

<tr>
	<td class="file medium" data-value="OtpInput.tsx"><a href="OtpInput.tsx.html">OtpInput.tsx</a></td>
	<td data-value="78.72" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 78%"></div><div class="cover-empty" style="width: 22%"></div></div>
	</td>
	<td data-value="78.72" class="pct medium">78.72%</td>
	<td data-value="47" class="abs medium">37/47</td>
	<td data-value="68" class="pct medium">68%</td>
	<td data-value="25" class="abs medium">17/25</td>
	<td data-value="76.92" class="pct medium">76.92%</td>
	<td data-value="13" class="abs medium">10/13</td>
	<td data-value="78.72" class="pct medium">78.72%</td>
	<td data-value="47" class="abs medium">37/47</td>
	</tr>

<tr>
	<td class="file high" data-value="TextFieldEndAdornment.tsx"><a href="TextFieldEndAdornment.tsx.html">TextFieldEndAdornment.tsx</a></td>
	<td data-value="83.33" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 83%"></div><div class="cover-empty" style="width: 17%"></div></div>
	</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="12" class="abs high">10/12</td>
	<td data-value="70.58" class="pct medium">70.58%</td>
	<td data-value="17" class="abs medium">12/17</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="12" class="abs high">10/12</td>
	</tr>

<tr>
	<td class="file high" data-value="constants.ts"><a href="constants.ts.html">constants.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="10" class="abs high">10/10</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="10" class="abs high">10/10</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2025-09-20T12:52:29.060Z
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    