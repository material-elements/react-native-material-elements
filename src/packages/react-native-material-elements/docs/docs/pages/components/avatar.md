# Avatar

The `Avatar` component is a versatile component that displays an image with various styling options. It extends the properties of the React Native `Image` component and includes additional styling options for variations and size.

![Avatar Component](https://docs-assets.developer.apple.com/published/b38ef3054b1d61b2a8f936cd81814d10/components-image-view-intro~dark@2x.png)

Within an image view, you can stretch, scale, size to fit, or pin the image to a specific location. Image views are typically not interactive.

## Props

The `Avatar` component accepts all props from the React Native `Image` component, in addition to the following props:

| Property     | type                         | Description                                         | Usage                                                                                        |
| ------------ | ---------------------------- | --------------------------------------------------- | -------------------------------------------------------------------------------------------- |
| `variation?` | String                       | Defines the shape of the image (default: `square`). | Options include `square`, `rounded`, `rounded-sm`, `rounded-md`, `rounded-lg`, `rounded-xl`. |
| `size?`      | Number                       | Size of the image.                                  | Specifies the dimensions of the image.                                                       |
| `sx?`        | [Styles](/utils/base-styles) | Custom styles for the image.                        | Allows for additional styling of the image.                                                  |

## Example

```tsx
import React from 'react';
import {Avatar} from 'react-native-material-elements';

export const Ex1: React.FC = () => {
  return (
    <Avatar
      source={{uri: 'your-image-source'}}
      sx={{w: 100, h: 100, o: 0.9, bg: 'red', d: 'flex'}}
    />
  );
};
```
