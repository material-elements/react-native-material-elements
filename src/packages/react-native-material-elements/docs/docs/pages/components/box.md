# Box

The `Box` component is a versatile container element that allows for easy layout and styling in React Native applications.

![Box component](https://docs-assets.developer.apple.com/published/5bb27442fab3a3bddc6f5949f1f343d9/components-box-intro~dark@2x.png)

## Best practices

**Prefer keeping a box relatively small in comparison with its containing view**. As a box’s size gets close to the size of the containing window or screen, it becomes less effective at communicating the separation of grouped content, and it can crowd other content.

**Consider using padding and alignment to communicate additional grouping within a box**. A box’s border is a distinct visual element — adding nested boxes to define subgroups can make your interface feel busy and constrained.

## Props

| Property   | Type                         | Description                                                                      |
| ---------- | ---------------------------- | -------------------------------------------------------------------------------- |
| `sx`       | [Styles](/utils/base-styles) | Object containing style properties for the Box (uses BaseStyles from styleTypes) |
| `children` | ReactNode                    | React node(s) to be rendered inside the Box.                                     |

## Examples

```tsx
<Box sx={{f: 1, d: 'flex', items: 'center', content: 'center'}}>
  <Box
    sx={{
      w: 50,
      h: 50,
      bg: 'white',
      r: 10,
      sColor: 'red',
      sOffset: {width: 0, height: 1},
      sOpacity: 0.8,
      sRadius: 1,
    }}
  />
</Box>
```
