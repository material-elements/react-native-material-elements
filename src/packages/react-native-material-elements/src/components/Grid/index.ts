export * from './Grid';
export * from './FlatGird';
