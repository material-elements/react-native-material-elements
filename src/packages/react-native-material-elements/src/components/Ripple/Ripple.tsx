import React, { useImperativeHandle, useMemo, useState } from 'react';
import { Animated, Easing, LayoutRectangle, StyleSheet, useColorScheme, View, ViewStyle } from 'react-native';
import { generateUniqueId } from '../../utils';
import { RIPPLE_DURATION, RIPPLE_RADIUS, RIPPLE_SIZE } from './constants';
import { RippleInterface, RippleObject, RipplePosition, RippleProps, onRippleAnimationType } from './Ripple.types';

const lightRippleBackground = 'rgba(0, 0, 0, 0.147)';
const darkRippleBackground = 'rgba(211, 211, 211, 0.174)';

const rippleContainerDefaultStyles: ViewStyle = {
  ...StyleSheet.absoluteFillObject,
  backgroundColor: 'transparent',
  overflow: 'hidden',
};

const rippleDefaultStyles: ViewStyle = {
  width: RIPPLE_RADIUS * 2,
  height: RIPPLE_RADIUS * 2,
  borderRadius: RIPPLE_RADIUS,
  overflow: 'hidden',
  position: 'absolute',
};

export const Ripple = React.forwardRef<RippleInterface, RippleProps>(
  ({ rippleStyles, rippleAnimationStyles, rippleContainerStyles, ...props }, ref) => {
    const [ripples, setRipples] = useState<Array<RippleObject>>([]);
    const colorScheme = useColorScheme();
    const isDarkMode = colorScheme === 'dark';

    const onRippleAnimationHandler: onRippleAnimationType = (animation, callBack) => {
      animation.start(callBack);
    };

    const onRippleAnimationEnd = () => {
      setRipples(prevState => prevState.slice(1));
    };

    const createRippleFromPosition = function (position: RipplePosition, elementLayoutRectangle: LayoutRectangle) {
      const { width, height } = elementLayoutRectangle;

      const positionCoordinates: Record<RipplePosition, [number, number]> = {
        center: [width / 2, height / 2],
        topLeft: [0, 0],
        topRight: [width, 0],
        bottomLeft: [0, height],
        bottomRight: [width, height],
      };

      if (position in positionCoordinates) {
        const [x, y] = positionCoordinates[position];
        startRipple(x, y);
      }
    };

    const startRipple = function (positionX: number, positionY: number) {
      let ripple = {
        progress: new Animated.Value(0),
        positionX: positionX - RIPPLE_SIZE / 2,
        positionY: positionY - RIPPLE_SIZE / 2,
        radius: RIPPLE_RADIUS,
        id: generateUniqueId(),
      };

      let animation = Animated.timing(ripple.progress, {
        toValue: 1,
        easing: Easing.out(Easing.ease),
        duration: RIPPLE_DURATION,
        useNativeDriver: true,
      });

      onRippleAnimationHandler(animation, onRippleAnimationEnd);
      setRipples(prevState => prevState.concat(ripple));
    };

    const animatedRippleStyles = useMemo(() => {
      const backgroundRippleColor = isDarkMode ? darkRippleBackground : lightRippleBackground;

      if (rippleStyles) {
        return { backgroundColor: backgroundRippleColor, ...rippleDefaultStyles, ...rippleStyles };
      } else {
        return { backgroundColor: backgroundRippleColor, ...rippleDefaultStyles };
      }
    }, [rippleStyles, isDarkMode]);

    const renderRipple = function (item: RippleObject) {
      let rippleAnimationDefaultStyles: ViewStyle = {
        top: item.positionY,
        left: item.positionX,
        transform: [{ scale: item.progress.interpolate({ inputRange: [0, 1], outputRange: [0.5, RIPPLE_SIZE * 2] }) }],
        opacity: item.progress.interpolate({
          inputRange: [0, 1],
          outputRange: [1, 0],
        }),
      };

      return (
        <Animated.View
          style={StyleSheet.flatten([
            animatedRippleStyles,
            [
              rippleAnimationStyles
                ? { ...rippleAnimationDefaultStyles, ...rippleAnimationStyles }
                : rippleAnimationDefaultStyles,
            ],
          ])}
          key={item.id}
        />
      );
    };

    useImperativeHandle(
      ref,
      () => {
        return {
          startRipple,
          createRippleFromPosition,
        } as RippleInterface;
      },
      // eslint-disable-next-line react-hooks/exhaustive-deps
      [],
    );

    return (
      <View style={[rippleContainerDefaultStyles, rippleContainerStyles]} {...props}>
        {ripples.map(item => renderRipple(item))}
      </View>
    );
  },
);

Ripple.displayName = 'Ripple';
