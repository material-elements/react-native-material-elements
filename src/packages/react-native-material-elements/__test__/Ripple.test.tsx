import React from 'react';
import { Ripple } from '../src/components/Ripple';
import { act, render } from './test-utils';
import { Animated } from 'react-native';

describe('Ripple effect', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly with default props', () => {
    const { toJSON } = render(<Ripple />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly', () => {
    const { getByTestId } = render(<Ripple testID="ripple" />);
    expect(getByTestId('ripple')).toBeTruthy();
  });
});
