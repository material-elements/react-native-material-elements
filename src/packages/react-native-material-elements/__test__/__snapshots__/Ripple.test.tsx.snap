// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ripple effect should render correctly with default props 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "transparent",
        "bottom": 0,
        "left": 0,
        "overflow": "hidden",
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      undefined,
    ]
  }
/>
`;
